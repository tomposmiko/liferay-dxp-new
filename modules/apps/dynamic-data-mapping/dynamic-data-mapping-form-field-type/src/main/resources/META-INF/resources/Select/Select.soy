{namespace Select}

{template .render}
	{@param name: string}
	{@param spritemap: string}
	{@param value: ?}
	{@param? _handleExpandedChanged: any}
	{@param? _handleItemClicked: any}
	{@param? _handleLabelClosed: any}
	{@param? dataSourceType: string}
	{@param? displayErrors: bool}
	{@param? errorMessage: string}
	{@param? expanded: bool}
	{@param? id: string}
	{@param? label: string}
	{@param? multiple: bool}
	{@param? options: list<?>}
	{@param? predefinedValue: ?}
	{@param? readOnly: bool}
	{@param? repeatable: bool}
	{@param? required: bool}
	{@param? showLabel: bool}
	{@param? tip: string}
	{@param? valid: bool}
	{@param? visible: bool}

	{let $displayValues: length($value) > 0 ? $value : $predefinedValue and length($predefinedValue) > 0 ? $predefinedValue : [] /}

	{call FieldBase.render}
		{param contentRenderer kind="html"}
			{call .content}
				{param _handleExpandedChanged: $_handleExpandedChanged /}
				{param _handleItemClicked: $_handleItemClicked /}
				{param _handleLabelClosed: $_handleLabelClosed /}
				{param dataSourceType: $dataSourceType /}
				{param displayValues: $displayValues /}
				{param expanded: $expanded /}
				{param id: $id /}
				{param multiple: $multiple /}
				{param name: $name /}
				{param options: $options /}
				{param readOnly: $readOnly /}
				{param spritemap: $spritemap /}
				{param value: $value /}
			{/call}
		{/param}
		{param displayErrors: $displayErrors /}
		{param errorMessage: $errorMessage /}
		{param label: $label /}
		{param name: $name /}
		{param readOnly: $readOnly /}
		{param ref: 'base' /}
		{param repeatable: $repeatable /}
		{param required: $required /}
		{param showLabel: $showLabel /}
		{param spritemap: $spritemap /}
		{param tip: $tip /}
		{param valid: $valid /}
		{param visible: $visible /}
	{/call}
{/template}

{template .content}
	{@param displayValues: ?}
	{@param name: string}
	{@param spritemap: string}
	{@param value: list<string>}
	{@param? _handleExpandedChanged: any}
	{@param? _handleItemClicked: any}
	{@param? _handleLabelClosed: any}
	{@param? expanded: bool}
	{@param? id: string}
	{@param? multiple: bool}
	{@param? options: list<?>}
	{@param? readOnly: bool}

	{if $options}
		{let $trigger kind="html"}
			{if not $readOnly}
				{call .hidden_select}
					{param displayValues: $displayValues /}
					{param multiple: $multiple /}
					{param name: $name /}
					{param options: $options /}
				{/call}
			{/if}

			<div class="form-builder-select-field input-group-container">
				{let $classes kind="text"}
					form-control results-chosen select-field-trigger

					{if $multiple}
						{sp}multiple-label-list
					{/if}
				{/let}

				{let $attributes kind="attributes"}
					class="{$classes}"

					{if $id}
						id="{$id}"
					{/if}

					{if $readOnly}
						disabled
					{/if}
				{/let}

				<ul {$attributes}>
					{foreach $currentValue in $value}
						{foreach $option in $options}
							{if $currentValue == $option.value}
								{if $multiple}
									<li>
										{call ClayLabel.render}
											{param closeable: true /}
											{param elementClasses: 'ddm-select-option-label' /}
											{param data: [
												'value': $option.value
											] /}
											{param events: [
												'close': $_handleLabelClosed
											]/}
											{param label: $option.label /}
											{param spritemap: $spritemap /}
										{/call}
									</li>
								{else}
									<div class="option-selected">{$option.label}</div>
								{/if}
							{/if}
						{/foreach}

						{ifempty}
							<div class="option-selected option-selected-placeholder">
								{if $multiple}
									{msg desc=""}choose-options{/msg}
								{else}
									{msg desc=""}choose-an-option{/msg}
								{/if}
							</div>
					{/foreach}

					<a class="select-arrow-down-container" href="javascript:;">
						{if $spritemap}
							{call ClayIcon.render}
								{param spritemap: $spritemap /}
								{param symbol: 'caret-double' /}
							{/call}
						{/if}
					</a>
				</ul>
			</div>
		{/let}

		{call ClayDropdownBase.render}
			{param disabled: $readOnly /}
			{param events: [
				'expandedChanged': $_handleExpandedChanged,
				'itemClicked': $_handleItemClicked
			]/}
			{param elementClasses: 'ddm-btn-full ddm-select-dropdown' /}
			{param expanded: $expanded /}
			{param items: $options /}
			{param label: $trigger /}
			{param ref: 'dropdown' /}
			{param searchable: length($options) > 6 ? true : false /}
			{param spritemap: $spritemap /}
			{param useSearchableWithForm: false /}
		{/call}
	{/if}
{/template}

/**
 * Prints hidden select tag.
 */
{template .hidden_select}
	{@param displayValues: ?}
	{@param name: string}
	{@param options: list<?>}
	{@param? multiple: bool}

	<select class="form-control hide" id="{$name}" name="{$name}" {if $multiple}multiple size="{length($options)}"{/if}>
		<option disabled {if length($displayValues) == 0}selected{/if} value=""></option>

		{foreach $option in $options}
			{call .select_hidden_options}
				{param option: $option /}
				{param values: $displayValues /}
			{/call}
		{/foreach}
	</select>
{/template}

/**
 * Prints the DDM form hidden select options.
 */
{template .select_hidden_options}
	{@param option: [label: string, value: string]}
	{@param values: ?}

	{let $selected kind="attributes"}
		{foreach $currentValue in $values}
			{if $currentValue == $option.value}
				selected
			{/if}
		{/foreach}
	{/let}

	<option {$selected} value="{$option.value}">{$option.label}</option>
{/template}