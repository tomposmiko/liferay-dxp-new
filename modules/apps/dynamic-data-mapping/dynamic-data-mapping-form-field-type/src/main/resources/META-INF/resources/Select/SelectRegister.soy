{namespace SelectRegister}

{deltemplate PageRenderer.RegisterFieldType variant="'select'"}
	{call Select.render data="all" /}
{/deltemplate}