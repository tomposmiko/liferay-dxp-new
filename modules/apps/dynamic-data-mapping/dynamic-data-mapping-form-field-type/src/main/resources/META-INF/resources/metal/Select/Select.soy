{namespace Select}

{template .render}
	{@param strings: [chooseAnOption: string]}
	{@param? _handleClick: any}
	{@param? _handleItemClicked: any}
	{@param? disableEmptyOption: bool}
	{@param? fixedOptions: list<[
		checked: bool,
		disabled: bool,
		id: string,
		inline: bool,
		label: string,
		name: string,
		showLabel: bool,
		value: string
	]>}
	{@param? id: string}
	{@param? label: string}
	{@param? open: bool}
	{@param? options: list<[
		checked: bool,
		disabled: bool,
		id: string,
		inline: bool,
		label: string,
		name: string,
		showLabel: bool,
		value: string
	]>}
	{@param? placeholder: string}
	{@param? predefinedValue: string}
	{@param? readOnly: bool}
	{@param? repeatable: bool}
	{@param? required: bool}
	{@param? selectedLabel: string}
	{@param? showLabel: bool}
	{@param? spritemap: string}
	{@param? tip: string}
	{@param? value: string}

	{call FieldBase.render}
		{param contentRenderer kind="html"}
			{call .content}
				{param _handleClick: $_handleClick /}
				{param _handleItemClicked: $_handleItemClicked /}
				{param disableEmptyOption: $disableEmptyOption /}
				{param fixedOptions: $fixedOptions /}
				{param id: $id /}
				{param open: $open /}
				{param options: $options /}
				{param placeholder: $placeholder /}
				{param predefinedValue: $predefinedValue /}
				{param readOnly: $readOnly /}
				{param selectedLabel: $selectedLabel /}
				{param spritemap: $spritemap /}
				{param strings: $strings /}
				{param value: $value /}
			{/call}
		{/param}

		{param id: $id /}
		{param label: $label /}
		{param repeatable: $repeatable /}
		{param required: $required /}
		{param showLabel: $showLabel /}
		{param spritemap: $spritemap /}
		{param tip: $tip /}
	{/call}
{/template}

{template .content}
	{@param strings: [chooseAnOption: string]}
	{@param? _handleItemClicked: any}
	{@param? _handleClick: any}
	{@param? disableEmptyOption: bool}
	{@param? fixedOptions: list<[
		checked: bool,
		disabled: bool,
		id: string,
		inline: bool,
		label: string,
		name: string,
		showLabel: bool,
		value: string
	]>}
	{@param? id: string}
	{@param? open: bool}
	{@param? options: list<[
		checked: bool,
		disabled: bool,
		id: string,
		inline: bool,
		label: string,
		name: string,
		showLabel: bool,
		value: string
	]>}
	{@param? placeholder: string}
	{@param? readOnly: bool}
	{@param? selectedLabel: string}
	{@param? spritemap: string}

	<div class="form-builder-select-field input-group-container">
		{let $attributes kind="attributes"}
			class="form-control select-field-trigger"

			{if $id}
				id="{$id}"
			{/if}

			{if $readOnly}
				disabled
			{/if}

			data-onclick="{$_handleClick}"
		{/let}

		<div {$attributes}>
			{if $selectedLabel}
				<div class="option-selected">{$selectedLabel}</div>
			{else}
				<div class="option-selected option-selected-placeholder">{$placeholder}</div>
			{/if}

			<a class="select-arrow-down-container" href="javascript:;">
				{if $spritemap}
					{call ClayIcon.render}
						{param spritemap: $spritemap /}
						{param symbol: 'caret-double' /}
					{/call}
				{/if}
			</a>
		</div>
	</div>

	{call .listItems}
		{param _handleItemClicked: $_handleItemClicked /}
		{param disableEmptyOption: $disableEmptyOption /}
		{param fixedOptions: $fixedOptions /}
		{param open: $open /}
		{param options: $options /}
		{param placeholder: $placeholder /}
		{param strings: $strings /}
	{/call}
{/template}

{template .listItems}
	{@param strings: [chooseAnOption: string]}
	{@param? _handleItemClicked: any}
	{@param? disableEmptyOption: bool}
	{@param? fixedOptions: list<[
		checked: bool,
		disabled: bool,
		id: string,
		inline: bool,
		label: string,
		name: string,
		showLabel: bool,
		value: string
	]>}
	{@param? open: bool}
	{@param? options: list<[
		checked: bool,
		disabled: bool,
		id: string,
		inline: bool,
		label: string,
		name: string,
		showLabel: bool,
		value: string
	]>}
	{@param? placeholder: string}

	{let $attributes kind="attributes"}
		class="dropdown-menu

		{if $open}
			{sp}dropdown-opened
		{/if}
		"
		data-onclick="{$_handleItemClicked}"

	{/let}

	{if $options}
		<div {$attributes}>
			<ul class="list-unstyled">
				{if not $disableEmptyOption}
					<li class="dropdown-item" data-option-value="">{$placeholder or $strings.chooseAnOption}</li>
				{/if}

				{if length($options) > 0}
					{foreach $option in $options}
						<li class="dropdown-item" data-option-value="{$option.value}">{$option.label}</li>
					{/foreach}
				{/if}

				{if $fixedOptions and length($fixedOptions) > 0}
					<li class="dropdown-divider"></li>

					{foreach $fixedOption in $fixedOptions}
						<li class="dropdown-item"  data-option-value="{$fixedOption.value}">{$fixedOption.label}</li>
					{/foreach}
				{/if}
			</ul>
		</div>
	{/if}
{/template}