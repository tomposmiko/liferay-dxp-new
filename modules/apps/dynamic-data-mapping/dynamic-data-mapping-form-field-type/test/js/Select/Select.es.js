/**
 * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
 *
 * This library is free software; you can redistribute it and/or modify it under
 * the terms of the GNU Lesser General Public License as published by the Free
 * Software Foundation; either version 2.1 of the License, or (at your option)
 * any later version.
 *
 * This library is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
 * details.
 */

import Select from '../../../src/main/resources/META-INF/resources/Select/Select.es';

let component;
const spritemap = 'icons.svg';

describe('Select', () => {
	beforeEach(() => {
		jest.useFakeTimers();
	});

	afterEach(() => {
		if (component) {
			component.dispose();
		}
	});

	it('is not editable', () => {
		component = new Select({
			readOnly: false,
			spritemap
		});

		expect(component).toMatchSnapshot();
	});

	it('has a helptext', () => {
		component = new Select({
			spritemap,
			tip: 'Type something'
		});

		expect(component).toMatchSnapshot();
	});

	it('has an id', () => {
		component = new Select({
			id: 'ID',
			spritemap
		});

		expect(component).toMatchSnapshot();
	});

	it('renders options', () => {
		component = new Select({
			options: [
				{
					checked: false,
					disabled: false,
					id: 'id',
					inline: false,
					label: 'label',
					name: 'name',
					showLabel: true,
					value: 'item'
				}
			],
			spritemap
		});

		expect(component).toMatchSnapshot();
	});

	it('renders no options when options come empty', () => {
		component = new Select({
			options: [],
			spritemap
		});

		expect(component).toMatchSnapshot();
	});

	it('has a label', () => {
		component = new Select({
			label: 'label',
			spritemap
		});

		expect(component).toMatchSnapshot();
	});

	it('is closed by default', () => {
		component = new Select({
			open: false,
			spritemap
		});

		expect(component).toMatchSnapshot();
	});

	it("has class dropdown-opened when it's opened", () => {
		component = new Select({
			open: true,
			spritemap
		});

		expect(component).toMatchSnapshot();
	});

	it('has a placeholder', () => {
		component = new Select({
			placeholder: 'Placeholder',
			spritemap
		});

		expect(component).toMatchSnapshot();
	});

	it('has a predefinedValue', () => {
		component = new Select({
			predefinedValue: ['Select'],
			spritemap
		});

		expect(component).toMatchSnapshot();
	});

	it('is not required', () => {
		component = new Select({
			required: false,
			spritemap
		});

		expect(component).toMatchSnapshot();
	});

	it('puts an asterisk when field is required', () => {
		component = new Select({
			label: 'This is the label',
			required: true,
			spritemap
		});

		expect(component).toMatchSnapshot();
	});

	it('renders Label if showLabel is true', () => {
		component = new Select({
			label: 'text',
			showLabel: true,
			spritemap
		});

		expect(component).toMatchSnapshot();
	});

	it('has a spritemap', () => {
		component = new Select({
			spritemap
		});

		expect(component).toMatchSnapshot();
	});

	it('has a value', () => {
		component = new Select({
			spritemap,
			value: ['value']
		});

		expect(component).toMatchSnapshot();
	});

	it('has a key', () => {
		component = new Select({
			key: 'key',
			spritemap
		});

		expect(component).toMatchSnapshot();
	});

	it('emits a field edit event when an item is selected', () => {
		const handleFieldEdited = jest.fn();

		const events = {fieldEdited: handleFieldEdited};

		jest.useFakeTimers();

		component = new Select({
			dataSourceType: 'manual',
			events,
			options: [
				{
					checked: false,
					disabled: false,
					id: 'id',
					inline: false,
					label: 'label',
					name: 'name',
					showLabel: true,
					value: 'item'
				}
			],
			spritemap
		});

		const spy = jest.spyOn(component, 'emit');

		jest.runAllTimers();

		component._handleItemClicked({
			data: {
				item: {
					value: 'Liferay'
				}
			},
			preventDefault: () => 0
		});

		expect(spy).toHaveBeenCalled();
	});

	it('renders the dropdown with search when there are more than six options', () => {
		component = new Select({
			dataSourceType: 'manual',
			options: [
				{
					label: 'label',
					name: 'name',
					value: 'item'
				},
				{
					label: 'label',
					name: 'name',
					value: 'item'
				},
				{
					label: 'label',
					name: 'name',
					value: 'item'
				},
				{
					label: 'label',
					name: 'name',
					value: 'item'
				},
				{
					label: 'label',
					name: 'name',
					value: 'item'
				},
				{
					label: 'label',
					name: 'name',
					value: 'item'
				},
				{
					label: 'label',
					name: 'name',
					value: 'item'
				}
			],
			spritemap
		});

		expect(component).toMatchSnapshot();
	});
});
