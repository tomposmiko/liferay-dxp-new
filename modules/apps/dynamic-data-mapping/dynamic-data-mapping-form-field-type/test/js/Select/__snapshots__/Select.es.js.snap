// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select should be closed by default 1`] = `
<div class="form-group">
  <div class="dropdown ddm-btn-full ddm-select-dropdown"><button aria-expanded="false" aria-haspopup="true" class="dropdown-toggle btn btn-unstyled" ref="triggerButton" type="button">
      <div class="form-builder-select-field input-group-container">
        <div class="form-control select-field-trigger">
          <div class="option-selected option-selected-placeholder">choose-an-option</div><a class="select-arrow-down-container" href="javascript:;"><svg class="lexicon-icon lexicon-icon-caret-double" focusable="false" role="presentation">
              <use xlink:href="icons.svg#caret-double"></use>
            </svg></a>
        </div>
      </div>
    </button></div>
</div>
<div class="dropdown-menu" ref="menu">
  <ul class="list-unstyled">
    <li tabindex="-1"><button class="btn dropdown-item btn-sm btn-link" aria-label="choose-an-option" type="button">choose-an-option</button></li>
  </ul>
</div>
`;

exports[`Select should be not edidable 1`] = `
<div class="form-group">
  <div class="dropdown ddm-btn-full ddm-select-dropdown"><button aria-expanded="false" aria-haspopup="true" class="dropdown-toggle btn btn-unstyled" ref="triggerButton" type="button">
      <div class="form-builder-select-field input-group-container">
        <div class="form-control select-field-trigger">
          <div class="option-selected option-selected-placeholder">choose-an-option</div><a class="select-arrow-down-container" href="javascript:;"><svg class="lexicon-icon lexicon-icon-caret-double" focusable="false" role="presentation">
              <use xlink:href="icons.svg#caret-double"></use>
            </svg></a>
        </div>
      </div>
    </button></div>
</div>
<div class="dropdown-menu" ref="menu">
  <ul class="list-unstyled">
    <li tabindex="-1"><button class="btn dropdown-item btn-sm btn-link" aria-label="choose-an-option" type="button">choose-an-option</button></li>
  </ul>
</div>
`;

exports[`Select should have a helptext 1`] = `
<div class="form-group">
  <div class="dropdown ddm-btn-full ddm-select-dropdown"><button aria-expanded="false" aria-haspopup="true" class="dropdown-toggle btn btn-unstyled" ref="triggerButton" type="button">
      <div class="form-builder-select-field input-group-container">
        <div class="form-control select-field-trigger">
          <div class="option-selected option-selected-placeholder">choose-an-option</div><a class="select-arrow-down-container" href="javascript:;"><svg class="lexicon-icon lexicon-icon-caret-double" focusable="false" role="presentation">
              <use xlink:href="icons.svg#caret-double"></use>
            </svg></a>
        </div>
      </div>
    </button></div><span class="form-text">Type something</span>
</div>
<div class="dropdown-menu" ref="menu">
  <ul class="list-unstyled">
    <li tabindex="-1"><button class="btn dropdown-item btn-sm btn-link" aria-label="choose-an-option" type="button">choose-an-option</button></li>
  </ul>
</div>
`;

exports[`Select should have a key 1`] = `
<div class="form-group">
  <div class="dropdown ddm-btn-full ddm-select-dropdown"><button aria-expanded="false" aria-haspopup="true" class="dropdown-toggle btn btn-unstyled" ref="triggerButton" type="button">
      <div class="form-builder-select-field input-group-container">
        <div class="form-control select-field-trigger">
          <div class="option-selected option-selected-placeholder">choose-an-option</div><a class="select-arrow-down-container" href="javascript:;"><svg class="lexicon-icon lexicon-icon-caret-double" focusable="false" role="presentation">
              <use xlink:href="icons.svg#caret-double"></use>
            </svg></a>
        </div>
      </div>
    </button></div>
</div>
<div class="dropdown-menu" ref="menu">
  <ul class="list-unstyled">
    <li tabindex="-1"><button class="btn dropdown-item btn-sm btn-link" aria-label="choose-an-option" type="button">choose-an-option</button></li>
  </ul>
</div>
`;

exports[`Select should have a label 1`] = `
<div class="form-group"><label>label </label>
  <div class="dropdown ddm-btn-full ddm-select-dropdown"><button aria-expanded="false" aria-haspopup="true" class="dropdown-toggle btn btn-unstyled" ref="triggerButton" type="button">
      <div class="form-builder-select-field input-group-container">
        <div class="form-control select-field-trigger">
          <div class="option-selected option-selected-placeholder">choose-an-option</div><a class="select-arrow-down-container" href="javascript:;"><svg class="lexicon-icon lexicon-icon-caret-double" focusable="false" role="presentation">
              <use xlink:href="icons.svg#caret-double"></use>
            </svg></a>
        </div>
      </div>
    </button></div>
</div>
<div class="dropdown-menu" ref="menu">
  <ul class="list-unstyled">
    <li tabindex="-1"><button class="btn dropdown-item btn-sm btn-link" aria-label="choose-an-option" type="button">choose-an-option</button></li>
  </ul>
</div>
`;

exports[`Select should have a placeholder 1`] = `
<div class="form-group">
  <div class="dropdown ddm-btn-full ddm-select-dropdown"><button aria-expanded="false" aria-haspopup="true" class="dropdown-toggle btn btn-unstyled" ref="triggerButton" type="button">
      <div class="form-builder-select-field input-group-container">
        <div class="form-control select-field-trigger">
          <div class="option-selected option-selected-placeholder">choose-an-option</div><a class="select-arrow-down-container" href="javascript:;"><svg class="lexicon-icon lexicon-icon-caret-double" focusable="false" role="presentation">
              <use xlink:href="icons.svg#caret-double"></use>
            </svg></a>
        </div>
      </div>
    </button></div>
</div>
<div class="dropdown-menu" ref="menu">
  <ul class="list-unstyled">
    <li tabindex="-1"><button class="btn dropdown-item btn-sm btn-link" aria-label="choose-an-option" type="button">choose-an-option</button></li>
  </ul>
</div>
`;

exports[`Select should have a predefinedValue 1`] = `
<div class="form-group">
  <div class="dropdown ddm-btn-full ddm-select-dropdown"><button aria-expanded="false" aria-haspopup="true" class="dropdown-toggle btn btn-unstyled" ref="triggerButton" type="button">
      <div class="form-builder-select-field input-group-container">
        <div class="form-control select-field-trigger">
          <div class="option-selected option-selected-placeholder">choose-an-option</div><a class="select-arrow-down-container" href="javascript:;"><svg class="lexicon-icon lexicon-icon-caret-double" focusable="false" role="presentation">
              <use xlink:href="icons.svg#caret-double"></use>
            </svg></a>
        </div>
      </div>
    </button></div>
</div>
<div class="dropdown-menu" ref="menu">
  <ul class="list-unstyled">
    <li tabindex="-1"><button class="btn dropdown-item btn-sm btn-link" aria-label="choose-an-option" type="button">choose-an-option</button></li>
  </ul>
</div>
`;

exports[`Select should have a spritemap 1`] = `
<div class="form-group">
  <div class="dropdown ddm-btn-full ddm-select-dropdown"><button aria-expanded="false" aria-haspopup="true" class="dropdown-toggle btn btn-unstyled" ref="triggerButton" type="button">
      <div class="form-builder-select-field input-group-container">
        <div class="form-control select-field-trigger">
          <div class="option-selected option-selected-placeholder">choose-an-option</div><a class="select-arrow-down-container" href="javascript:;"><svg class="lexicon-icon lexicon-icon-caret-double" focusable="false" role="presentation">
              <use xlink:href="icons.svg#caret-double"></use>
            </svg></a>
        </div>
      </div>
    </button></div>
</div>
<div class="dropdown-menu" ref="menu">
  <ul class="list-unstyled">
    <li tabindex="-1"><button class="btn dropdown-item btn-sm btn-link" aria-label="choose-an-option" type="button">choose-an-option</button></li>
  </ul>
</div>
`;

exports[`Select should have a value 1`] = `
<div class="form-group">
  <div class="dropdown ddm-btn-full ddm-select-dropdown"><button aria-expanded="false" aria-haspopup="true" class="dropdown-toggle btn btn-unstyled" ref="triggerButton" type="button">
      <div class="form-builder-select-field input-group-container">
        <div class="form-control select-field-trigger">
          <div class="option-selected option-selected-placeholder">choose-an-option</div><a class="select-arrow-down-container" href="javascript:;"><svg class="lexicon-icon lexicon-icon-caret-double" focusable="false" role="presentation">
              <use xlink:href="icons.svg#caret-double"></use>
            </svg></a>
        </div>
      </div>
    </button></div>
</div>
<div class="dropdown-menu" ref="menu">
  <ul class="list-unstyled">
    <li tabindex="-1"><button class="btn dropdown-item btn-sm btn-link" aria-label="choose-an-option" type="button">choose-an-option</button></li>
  </ul>
</div>
`;

exports[`Select should have an id 1`] = `
<div class="form-group">
  <div class="dropdown ddm-btn-full ddm-select-dropdown"><button aria-expanded="false" aria-haspopup="true" class="dropdown-toggle btn btn-unstyled" ref="triggerButton" type="button">
      <div class="form-builder-select-field input-group-container">
        <div class="form-control select-field-trigger" id="ID">
          <div class="option-selected option-selected-placeholder">choose-an-option</div><a class="select-arrow-down-container" href="javascript:;"><svg class="lexicon-icon lexicon-icon-caret-double" focusable="false" role="presentation">
              <use xlink:href="icons.svg#caret-double"></use>
            </svg></a>
        </div>
      </div>
    </button></div>
</div>
<div class="dropdown-menu" ref="menu">
  <ul class="list-unstyled">
    <li tabindex="-1"><button class="btn dropdown-item btn-sm btn-link" aria-label="choose-an-option" type="button">choose-an-option</button></li>
  </ul>
</div>
`;

exports[`Select should have class dropdown-opened when it's opened 1`] = `
<div class="form-group">
  <div class="dropdown ddm-btn-full ddm-select-dropdown"><button aria-expanded="false" aria-haspopup="true" class="dropdown-toggle btn btn-unstyled" ref="triggerButton" type="button">
      <div class="form-builder-select-field input-group-container">
        <div class="form-control select-field-trigger">
          <div class="option-selected option-selected-placeholder">choose-an-option</div><a class="select-arrow-down-container" href="javascript:;"><svg class="lexicon-icon lexicon-icon-caret-double" focusable="false" role="presentation">
              <use xlink:href="icons.svg#caret-double"></use>
            </svg></a>
        </div>
      </div>
    </button></div>
</div>
<div class="dropdown-menu" ref="menu">
  <ul class="list-unstyled">
    <li tabindex="-1"><button class="btn dropdown-item btn-sm btn-link" aria-label="choose-an-option" type="button">choose-an-option</button></li>
  </ul>
</div>
`;

exports[`Select should not be required 1`] = `
<div class="form-group">
  <div class="dropdown ddm-btn-full ddm-select-dropdown"><button aria-expanded="false" aria-haspopup="true" class="dropdown-toggle btn btn-unstyled" ref="triggerButton" type="button">
      <div class="form-builder-select-field input-group-container">
        <div class="form-control select-field-trigger">
          <div class="option-selected option-selected-placeholder">choose-an-option</div><a class="select-arrow-down-container" href="javascript:;"><svg class="lexicon-icon lexicon-icon-caret-double" focusable="false" role="presentation">
              <use xlink:href="icons.svg#caret-double"></use>
            </svg></a>
        </div>
      </div>
    </button></div>
</div>
<div class="dropdown-menu" ref="menu">
  <ul class="list-unstyled">
    <li tabindex="-1"><button class="btn dropdown-item btn-sm btn-link" aria-label="choose-an-option" type="button">choose-an-option</button></li>
  </ul>
</div>
`;

exports[`Select should put an asterisk when field is required 1`] = `
<div class="form-group"><label>This is the label <span class="reference-mark"><svg class="lexicon-icon lexicon-icon-asterisk" focusable="false" role="presentation">
        <use xlink:href="icons.svg#asterisk"></use>
      </svg></span></label>
  <div class="dropdown ddm-btn-full ddm-select-dropdown"><button aria-expanded="false" aria-haspopup="true" class="dropdown-toggle btn btn-unstyled" ref="triggerButton" type="button">
      <div class="form-builder-select-field input-group-container">
        <div class="form-control select-field-trigger">
          <div class="option-selected option-selected-placeholder">choose-an-option</div><a class="select-arrow-down-container" href="javascript:;"><svg class="lexicon-icon lexicon-icon-caret-double" focusable="false" role="presentation">
              <use xlink:href="icons.svg#caret-double"></use>
            </svg></a>
        </div>
      </div>
    </button></div>
</div>
<div class="dropdown-menu" ref="menu">
  <ul class="list-unstyled">
    <li tabindex="-1"><button class="btn dropdown-item btn-sm btn-link" aria-label="choose-an-option" type="button">choose-an-option</button></li>
  </ul>
</div>
`;

exports[`Select should render Label if showLabel is true 1`] = `
<div class="form-group"><label>text </label>
  <div class="dropdown ddm-btn-full ddm-select-dropdown"><button aria-expanded="false" aria-haspopup="true" class="dropdown-toggle btn btn-unstyled" ref="triggerButton" type="button">
      <div class="form-builder-select-field input-group-container">
        <div class="form-control select-field-trigger">
          <div class="option-selected option-selected-placeholder">choose-an-option</div><a class="select-arrow-down-container" href="javascript:;"><svg class="lexicon-icon lexicon-icon-caret-double" focusable="false" role="presentation">
              <use xlink:href="icons.svg#caret-double"></use>
            </svg></a>
        </div>
      </div>
    </button></div>
</div>
<div class="dropdown-menu" ref="menu">
  <ul class="list-unstyled">
    <li tabindex="-1"><button class="btn dropdown-item btn-sm btn-link" aria-label="choose-an-option" type="button">choose-an-option</button></li>
  </ul>
</div>
`;

exports[`Select should render no options when options come empty 1`] = `
<div class="form-group">
  <div class="dropdown ddm-btn-full ddm-select-dropdown"><button aria-expanded="false" aria-haspopup="true" class="dropdown-toggle btn btn-unstyled" ref="triggerButton" type="button">
      <div class="form-builder-select-field input-group-container">
        <div class="form-control select-field-trigger">
          <div class="option-selected option-selected-placeholder">choose-an-option</div><a class="select-arrow-down-container" href="javascript:;"><svg class="lexicon-icon lexicon-icon-caret-double" focusable="false" role="presentation">
              <use xlink:href="icons.svg#caret-double"></use>
            </svg></a>
        </div>
      </div>
    </button></div>
</div>
<div class="dropdown-menu" ref="menu">
  <ul class="list-unstyled">
    <li tabindex="-1"><button class="btn dropdown-item btn-sm btn-link" aria-label="choose-an-option" type="button">choose-an-option</button></li>
  </ul>
</div>
`;

exports[`Select should render options 1`] = `
<div class="form-group">
  <div class="dropdown ddm-btn-full ddm-select-dropdown"><button aria-expanded="false" aria-haspopup="true" class="dropdown-toggle btn btn-unstyled" ref="triggerButton" type="button">
      <div class="form-builder-select-field input-group-container">
        <div class="form-control select-field-trigger">
          <div class="option-selected option-selected-placeholder">choose-an-option</div><a class="select-arrow-down-container" href="javascript:;"><svg class="lexicon-icon lexicon-icon-caret-double" focusable="false" role="presentation">
              <use xlink:href="icons.svg#caret-double"></use>
            </svg></a>
        </div>
      </div>
    </button></div>
</div>
<div class="dropdown-menu" ref="menu">
  <ul class="list-unstyled">
    <li tabindex="-1"><button class="btn dropdown-item btn-sm btn-link" aria-label="choose-an-option" type="button">choose-an-option</button></li>
    <li tabindex="-1"><button class="btn dropdown-item btn-sm btn-link" aria-label="label" type="button">label</button></li>
  </ul>
</div>
`;

exports[`Select should render the dropdown with search when there are more than six options 1`] = `
<div class="form-group">
  <div class="dropdown ddm-btn-full ddm-select-dropdown"><button aria-expanded="false" aria-haspopup="true" class="dropdown-toggle btn btn-unstyled" ref="triggerButton" type="button">
      <div class="form-builder-select-field input-group-container">
        <div class="form-control select-field-trigger">
          <div class="option-selected option-selected-placeholder">choose-an-option</div><a class="select-arrow-down-container" href="javascript:;"><svg class="lexicon-icon lexicon-icon-caret-double" focusable="false" role="presentation">
              <use xlink:href="icons.svg#caret-double"></use>
            </svg></a>
        </div>
      </div>
    </button></div>
</div>
<div class="dropdown-menu" ref="menu">
  <form>
    <div class="dropdown-section">
      <div class="input-group input-group-sm">
        <div class="input-group-item"><input aria-label="Search for..." class="form-control input-group-inset input-group-inset-after" placeholder="Search for..." type="text" ref="searchInput"><span class="input-group-inset-item input-group-inset-item-after"><button class="btn btn-link" aria-label="search" type="button"><svg class="lexicon-icon lexicon-icon-search" focusable="false" role="presentation">
                <use xlink:href="icons.svg#search"></use>
              </svg></button></span></div>
      </div>
    </div>
  </form>
  <ul class="list-unstyled">
    <li tabindex="-1"><button class="btn dropdown-item btn-sm btn-link" aria-label="choose-an-option" type="button">choose-an-option</button></li>
    <li tabindex="-1"><button class="btn dropdown-item btn-sm btn-link" aria-label="label" type="button">label</button></li>
    <li tabindex="-1"><button class="btn dropdown-item btn-sm btn-link" aria-label="label" type="button">label</button></li>
    <li tabindex="-1"><button class="btn dropdown-item btn-sm btn-link" aria-label="label" type="button">label</button></li>
    <li tabindex="-1"><button class="btn dropdown-item btn-sm btn-link" aria-label="label" type="button">label</button></li>
    <li tabindex="-1"><button class="btn dropdown-item btn-sm btn-link" aria-label="label" type="button">label</button></li>
    <li tabindex="-1"><button class="btn dropdown-item btn-sm btn-link" aria-label="label" type="button">label</button></li>
    <li tabindex="-1"><button class="btn dropdown-item btn-sm btn-link" aria-label="label" type="button">label</button></li>
  </ul>
</div>
`;
